File: Leetcode/_206_Reverse_Linked_List.java
// link - https://leetcode.com/problems/reverse-linked-list/

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class _206_Reverse_Linked_List {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        while (curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    public ListNode reverseListRec(ListNode head) {
        return reverseListRecursive(null, head, null);
    }

    public ListNode reverseListRecursive(ListNode prev, ListNode curr, ListNode newHead) {
        if (curr == null) {
            newHead = prev;
            return newHead;
        }
        newHead = reverseListRecursive(curr, curr.next, newHead);
        curr.next = prev;
        return newHead;
    }
}

File: Leetcode/python/_206_Reverse_Linked_List.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    # Iterative - Time = O(n) - Space = O(1)
    # Its a linked list, just reverse the links and move head to last node
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        while head:
            temp = head.next
            head.next = prev
            prev = head
            head = temp
        return prev

    # Recursive - Time = O(n) - Space = O(n)
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def rec(cur, prev):
            if not cur:
                return prev
            else:
                next = cur.next
                cur.next = prev
                return rec(next, cur)

        return rec(head, None)
